### PR TITLE
round out local Makefile checks

### PR DESCRIPTION
**Pull request by geekosaur**
by collecting them and adding hlint. Other checks will follow, e.g. API checking once that lands.

Some other things might be added here later, and might lead to revision of various CI jobs. Consider this a start on addressing #10263 properly.

**Template B: This PR does not modify behaviour or interface**

*E.g. the PR only touches documentation or tests, does refactorings, etc.*

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Is this a PR that fixes CI? If so, it will need to be backported to older cabal release branches (ask maintainers for directions).

### CODE REVIEW

**Comment by geekosaur** [comment]
I did a once-over of the "quick checks", which are currently not covered; most of those, however, either require a pristine CI-like build environment (or docker) or require unreasonable dependencies (e.g. Sphinx). The one exception is the doctests, which I'm on the fence about.

There's also a question of how best to deal with `fourmolu`, considering that we require 0.12 and using 0.14 will cause failures either locally or in CI.

File: Makefile
Comment (ulysses4ever) [inline Makefile:56]: Why not do it as a part of this PR though?

**Review by ulysses4ever — APPROVED** [review]

File: Makefile
Comment (ulysses4ever) [reply Makefile:56]: Fair enough. I don't feel too strong about it. I just can't stand the look of Makefile when I found it about 90% bit-rooted sometime in 2021, I think. I feel like actions are good for setting up the environment (e.g. installing tools) but not very good at running stuff.

Comment (geekosaur) [reply Makefile:56]: We have a few examples of that one as well, but this is actually spread between the setup and run actions. In particular, there is a flag in the setup action that I'm confused about because it looks like it's about the run action.

Comment (geekosaur) [reply Makefile:56]: I am hoping this won't bitrot because the whole point is that people should use it before pushing to CI, so hopefully any problems will be reported early.

Comment (geekosaur) [reply Makefile:56]: Perhaps it's time for a quick-start guide. Which should emphasize this feature because it's designed to make their life easier: they don't have to wait for CI to catch things they can find and fix locally.

Comment (ulysses4ever) [reply Makefile:56]: That's a good idea.

**Review by ffaf1 — APPROVED** [review]

**Review by mpickering — APPROVED** [review]
Thanks @geekosaur :)

**Comment by geekosaur** [comment]
Once again merged as soon as "Validate Skip / Validate post job" went green and the other branch protection rules were satisfied (mostly this means "Quick checks", which despite the name take 7-10 minutes).

**Pull request merged**
